p2p/src/swarm/peerdb.rs

```diff
@@ -30,86 +30,119 @@
 #[derive(Debug)]
 pub struct PeerContext<T: NetworkingService> {
     /// Peer information
-    pub _info: types::PeerInfo<T>,
+    pub info: types::PeerInfo<T>,
+
+    /// Peer's active address, if known
+    pub address: Option<T::Address>,
+
+    /// Set of available addresses
+    pub addresses: HashSet<T::Address>,
 
     /// Peer score
     pub score: u32,
 }
 
-/// Peer address information
-enum PeerAddrInfo<T: NetworkingService> {
-    Raw {
-        /// Hashset of IPv4 addresses
-        ip4: HashSet<T::Address>,
+#[derive(Debug)]
+pub enum Peer<T: NetworkingService> {
+    /// Active peer
+    Active(PeerContext<T>),
+
+    /// Idle peer (`types::PeerInfo<t>` received)
+    Idle(PeerContext<T>),
+
+    /// Active/known peer that has been banned
+    Banned(Either<PeerContext<T>, VecDeque<T::Address>>),
 
-        /// Hashset of IPv6 addresses
-        ip6: HashSet<T::Address>,
-    },
+    /// Discovered peer (addresses have been received)
+    Discovered(VecDeque<T::Address>),
 }
 
-// TODO: improve how peers are stored, order by reputation?
+// TODO: store available peers into a binary heap
 pub struct PeerDb<T: NetworkingService> {
-    /// Hashmap for peer information
-    peers: HashMap<T::PeerId, PeerContext<T>>,
+    /// P2P configuration
+    p2p_config: Arc<config::P2pConfig>,
 
-    /// Hashmap of discovered peers we don't have an active connection with
-    discovered: HashMap<T::PeerId, PeerAddrInfo<T>>,
+    /// Set of peers known to `PeerDb`
+    peers: HashMap<T::PeerId, Peer<T>>,
+
+    /// Set of available (idle) peers
+    available: HashSet<T::PeerId>,
 
     /// Pending connections
-    pending: HashMap<T::Address, PeerAddrInfo<T>>,
+    pending: HashMap<T::Address, T::PeerId>,
 
     /// Banned peers
     banned: HashSet<T::PeerId>,
 }
 
 impl<T: NetworkingService> PeerDb<T> {
-    pub fn new() -> Self {
+    pub fn new(p2p_config: Arc<config::P2pConfig>) -> Self {
         Self {
-            peers: HashMap::new(),
-            discovered: HashMap::new(),
-            pending: HashMap::new(),
-            banned: HashSet::new(),
+            peers: Default::default(),
+            available: Default::default(),
+            pending: Default::default(),
+            banned: Default::default(),
+            p2p_config,
         }
     }
 
     /// Get the number of idle (available) peers
     pub fn idle_peer_count(&self) -> usize {
-        self.discovered.len()
+        self.available.len()
     }
 
-    /// Get socket address of the next best peer (TODO: in terms of peer score)
-    pub fn best_peer_addr(&mut self) -> Option<T::Address> {
-        // TODO: improve peer selection
-        let key = match self.discovered.keys().next() {
-            Some(key) => *key,
-            None => return None,
-        };
+    /// Get the number of active peers
+    pub fn active_peer_count(&self) -> usize {
+        self.peers
+            .len()
+            .saturating_sub(self.pending.len())
```

Comment:
I'm a big opponent of using saturating sub. It basically hides bugs. Is there a way we can do this in a better, consistent way that reveals errors?

Reply:
I believe this `saturating_sub` is basically equivalent to `max(0, peers_len - pending_len)`, but without conversion and underflow issues, it's a well defined operation. In fact, there is a lint in clippy that suggests `saturating_sub` as a better behaved alternative to the `max` expression above.

Reply:
If it's being as an alternative to max, I get it. That's OK. But I think in many cases developers use it just to get rid of underflow/overflow errors and its handling, which might lead to inconsistencies.

Reply:
To be more specific, I'm fine with strongly discouraging `T::saturating_add` and  `S::saturating_sub`, where `T` is any integer type and `S` is a signed integer type. I'd like to keep `U::saturating_sub` for unsigned types `U`, which is basically [monus](https://en.wikipedia.org/wiki/Monus).

Reply:
Guys, I'm still having trouble understanding this. Help me understand what's going on.

If the number of planets in the solar system is 8, and we want to know the number of non-red and non-blue planets in the system, we do `total_planet_count - total_red - total_blue`, so `8 - 1 - 1` which yields `6` (assuming earth is blue and mars is red). Good. Now, how does it help to make it `total_planet_count.saturating_sub(total_red).saturating_sub(total_blue)`? Yes, if all the inputs are correct, then this will still yield seven. But if you miscounted something (broke the invariants of your system), then the true result can be 5, or 3, or 0... but then you'll never catch anything less than zero (an underflow) because it's a saturating sub. Meaning, a bug will always give you wrong results.

How does using `max(0, total_planet_count - total_red)` help in this case, based on Lukas's assumption that this is what we're doing?

Reply:
The most robust way to calculate number of active peers is to iterate over the collection and count the number of active peers.

Reply:
I think I'm not making my point clear. I'll try again. Let's consider the equation (and please tell me if the equation is incorrect):

`all_peers = active + banned + available + pending`.

And if I understand correctly, all those types are orthogonal. Meaning, a peer can be only one of these at a time. Right?

In your variables that represents that model, you don't have active, and if I understand correctly, active is basically:

`active = all_peers - banned - available - pending`,

which is what you're doing in this subtraction chain. Am I getting this right? If yes, then this equation is an invariant. There's absolutely no way this will change unless something is broken. IF something is broken, then using saturating_sub will only cause hiding bugs.

The only way saturating_sub is acceptable is if those types aren't orthogonal.

Reply:
I don't mind changing it to `expect()` and removing `saturating_sub()`

Reply:
If my understanding is correct, then using expect is the right way to go. What do you think @iljakuklic ?